File: hw4/HW4S20/LTMatrix.cs
using System;
namespace HW4 {
    public class LTMatrix {
        private int[] arr;
        private int n;
        
        public LTMatrix(int N) {
            int numElements = (int) ((1+N)*N)/2;
            arr = new int[numElements];
            n = N;
        }

        public int Length() {
            return arr.Length;
        }

        public int Size() {
            return n;
        }

        public int[] OneDArray() {
            return arr;
        }

        public int TwoDToOneDIndex(int row, int column) {
            if (row < column) {
                throw new Exception("row should be >= column");
            } else {
                int index = (row*(row+1))/2 + column;
                return index;
            }
        }

        public void Set(int row, int column, int V) {
            if (row > (n - 1) && column > (n - 1)) {
                throw new Exception("out of bounds");
            } else{
                int index = this.TwoDToOneDIndex(row, column);
                arr[index] = V;
            }
        }

        public int Get(int row, int column) {
            if(row <= (n - 1) && column <= (n - 1) && row < column) {
                return 0;
            } else {
                int index = this.TwoDToOneDIndex(row, column);
                return arr[index];
            }
        }

        public static LTMatrix operator+ (LTMatrix lt1, LTMatrix lt2) {
            if (lt1.Size() != lt2.Size()) {
                throw new Exception("cannot add different sized matrices");
            } else {
                LTMatrix result = new LTMatrix(lt1.Size());
                for (int i = 0; i < lt1.Size(); i++) {
                    for (int j = 0; j < lt1.Size(); j++) {
                        int value = lt1.Get(i,j) + lt2.Get(i,j);
                        result.Set(i,j, value);
                    }
                }
                return result;
            }
        }
    }
}

File: hw4/HW4S20/ltmatrix.py
class ltMatrix:
    def __init__(self, N):
        self.arr = [None]*(((1+N) * N)//2)
        self.N = N
    
    def length(self):
        return len(self.arr)

    def size(self):
        return self.N

    def oneDArray(self):
        return self.arr

    def twoDToOneDIndex(self, row, column):
        if row < column:
            raise Exception('row should be >= column')
        else:
            index = (row*(row+1))//2 + column
            return index

    def set(self, row, column, V):
        if row > (self.N - 1) and column > (self.N - 1):
            raise Exception('out of bounds')
        else:
            index = self.twoDToOneDIndex(row, column)
            #print("row: %d column: %d index: %d", row, column, index)
            self.arr[index] = V
    
    def get(self, row, column):
        if row <= (self.N - 1) and column <= (self.N - 1) and row < column:
            return 0
        else:
            index = self.twoDToOneDIndex(row, column)
            return self.arr[index]

    def __add__(self, other):
        if not (other.N == self.N):
            raise Exception('cannot add different sized matrices')
        else:
            result = ltMatrix(self.N)
            for i in range(0,self.N):
                for j in range(0, self.N):
                    value = self.get(i,j) + other.get(i,j)
                    result.set(i,j,value)
            return result